src/libostree/ostree-repo.c

```diff
@@ -374,31 +386,47 @@ push_repo_lock (OstreeRepo          *self,
   repo_lock_info (lock, &info);
   g_debug ("Push lock: state=%s, depth=%u", info.name, info.len);
 
-  if (info.state == LOCK_EX)
+  int next_state;
+  guint *counter;
+  if (lock_type == OSTREE_REPO_LOCK_EXCLUSIVE)
     {
-      g_debug ("Repo already locked exclusively, extending stack");
-      g_queue_push_head (&lock->stack, GINT_TO_POINTER (LOCK_EX));
+      next_state = LOCK_EX;
+      counter = &(lock->exclusive);
     }
   else
     {
-      int next_state = (flags & LOCK_EX) ? LOCK_EX : LOCK_SH;
-      const char *next_state_name = (flags & LOCK_EX) ? "exclusive" : "shared";
+      next_state = LOCK_SH;
+      counter = &(lock->shared);
+    }
+
+  /* Check for overflow */
+  g_assert_cmpuint (*counter, <, G_MAXUINT);
+
+  if (info.state == LOCK_EX || info.state == next_state)
+    {
+      g_debug ("Repo already locked %s, extending stack", info.name);
+    }
+  else
+    {
+      const char *next_state_name = lock_state_name (next_state);
 
       g_debug ("Locking repo %s", next_state_name);
       if (!do_repo_lock (lock->fd, flags))
```

Comment:
Maybe worth an assertion somewhere round here that checks a lock is never ‘upgraded’ from exclusive to shared.

Reply:
:+1: